### ISSUE TITLE
MinBy / MaxBy behavior w.r.t key value types and nulls

### ISSUE THREAD

**Issue by LordJZ** [dotnet/runtime#53528]
I am testing the new MinBy and MaxBy methods. I find their behavior really weird with regards to nullability. Consider this piece of code:

```C#
> new string[0].MinBy(x => x + "")
null
> new string[0].MinBy(x => x.Length)
System.InvalidOperationException: Sequence contains no elements
> new int[0].MinBy(x => x + "")
0 // sic!!
> new int[0].MinBy(x => x)
System.InvalidOperationException: Sequence contains no elements
```

Here the MinBy method throws an exception, when the _key_ is value type, but not the return type is value type. I understand that this mirrors Min method.

It feels to me it would make a better API to instead throw based on whether the return type is a value type. In other words, change this to `if (value == null)`:
https://github.com/dotnet/runtime/blob/abd01f544049fe32aefda440efd98a7ff3e52583/src/libraries/System.Linq/src/System/Linq/Min.cs#L492

Consider this use case scenario:

```C#
 new string[0].OrderBy(x => x.Length).FirstOrDefault(); // returns null
```

Should this be refactored into MinBy?
```C#
new string[0].MinBy(x => x.Length); // currently throws, but should it?
```

I feel like it makes sense for both of these statements to return null:
```C#
new string[0].MinBy(x => x.Length); // currently throws, but should it?
new string[0].MinBy(x => x); // returns null
```
I feel like it makes sense for both of these statements to throw:
```C#
new int[0].MinBy(x => x); // currently throws
new int[0].MinBy(x => x + ""); // returns 0, but should it?
```

Bottom line: I think it's better for MinBy/MaxBy to mirror behavior of `.OrderBy[Descending](...).First[OrDefault]()` rather then mirror Min/Max.

**Comment by Joe4evr** [comment]
> ```cs
> > new int[0].MinBy(x => x + "")
> 0 // sic!!
> ```

🤣 
Looks like C# got its own case of `{} + []` now.

**Comment by eiriktsarpalis** [comment]
> Here the MinBy method throws an exception, when the key is value type, but not the return type is value type. I understand that this mirrors Min method.

You're right, this is incorrect behaviour and it should be fixed. Thanks for reporting it.


I've submitted a PR that addresses the issue. https://github.com/dotnet/runtime/pull/53544
